enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,467 @@
+
+New configuration and a reconcile loop for the
+`control-plane-machine-set-operator` will be introduced that will monitor
+resource usage of the control plane nodes and automatically modify the
+`ControlPlaneMachineSet` if scaling the nodes is determined to be required.
+
+The new configuration will specify if scaling up and down is enabled and allow
+configuring when to trigger scaling. It will also allow configuring which
+machine types are to be used for scaling, putting an upper and lower limit on
+possible automatic scaling.
+
+### Workflow Description
+
+1. The OpenShift adminstrator creates a valid
+   `control-plane-machine-set-autoscaling` `CR` in the `openshift-machine-api`
+   namespace (or the respective namespace `control-plane-machineset-operator` is
+   running in), to configure the automatic vertical scaling.
+
+### API Extensions
+
+This proposal requires a new custom resource to configure when and how
+automatically scale the control plane by modifying the control plane machinset.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The implementation for this feature, requires a data source to gather required
+data that is used to make scaling decisions.
+
+As every OpenShift cluster comes with the [metrics
+API](https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/)
+as well as
+[prometheus](https://docs.openshift.com/container-platform/4.13/monitoring/monitoring-overview.html)
+installed, both should be available as possible data sources.
+
+#### Using metrics API
+
+Current CPU and memory usage for nodes, as well as available resources on the
+nodes are gathered using the metrics API.
+
+To gather available resource information in this case, the default
+[Node](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#node-v1-core)
+API endpoint can be used.
+
+Using the [horizontal pod
+autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)
+as a reference, they are using the to gather performance statistics for nodes.
+
+Using the metrics API has the least requirements as it does not expect a running
+Prometheus monitoring stack. 
+
+On the other hand, data aggregation will be required using the metrics API, as
+decisions about scaling nodes up and down, must not be performed based on
+performance usage at a single point in time. Instead the operator will have to
+gather and store usage data for a certain period of time, to decide if scaling
+is required.
+
+As an example with a configured `timeWindow` (see
+[configuration](#control-plane-machineset-operator-configuration)) of 30 minutes
+for scaling up, the operator has to keep querying the metrics API and averaging
+values until enough data has been aggregated to know the load average for the
+last 30 minutes.
+
+If the metrics API returns it's own window of ~60 second aggregations, the
+operators will have to retrieve 30 values (one per minute), before any scaling
+decision should be made.
```

Comment:
Not really - I think this will be something that will be part of the implementation having to validate what works best - I don't really want to dictate this in the enhancement as e.g. a rolling average of X over time as this might not be the best.
Is leaving this 'out' and adding it later once an implementation is started to check what works best can be used to gather information or do we need something before this can be merged?

Reply:
You could document it as an open question and come back later once you've solved the issue for sure